Packages/com.unity.inputsystem/InputSystem/Editor/InputActionAsset/AddControlSchemePopup.cs

```diff
@@ -0,0 +1,367 @@
+        int m_RequirementsOptionsChoice;
+
+        InputActionAssetManager m_AssetManager;
+        InputActionWindowToolbar m_Toolbar;
+
+        bool m_SetFocus;
+        Vector2 m_SchemaNameLabelSize;
+        float m_RequirementHeights;
+        float m_ButtonsAndLabelsHeights;
+
+        static readonly GUIContent m_RequirementGUI = new GUIContent("Requirements:");
+        static readonly GUIContent m_AddControlSchemeGUI = new GUIContent("Add control scheme");
+        static readonly GUIContent m_SchemaNameGUI = new GUIContent("Scheme Name");
+        static readonly Vector2 s_DefaultSize = new Vector2(300, 200);
+        static readonly string[] choices = { "Optional", "Required" };
+        string m_OriginalName;
+        Action m_Apply;
+
+        public AddControlSchemePopup(InputActionAssetManager assetManager, InputActionWindowToolbar toolbar, Action apply)
+        {
+            m_AssetManager = assetManager;
+            m_Toolbar = toolbar;
+            m_SetFocus = true;
+            m_Apply = apply;
+        }
+
+        public void SetSchemaForEditing(string schemaName)
+        {
+            for (int i = 0; i < m_AssetManager.m_AssetObjectForEditing.m_ControlSchemes.Length; i++)
+            {
+                if (m_AssetManager.m_AssetObjectForEditing.m_ControlSchemes[i].name == schemaName)
+                {
+                    m_ControlSchemeIndex = i;
+                    break;
+                }
+            }
+            SetSchemaParametersFrom(schemaName);
+            m_OriginalName = schemaName;
+        }
+
+        public void DuplicateParametersFrom(string schemaName)
+        {
+            SetSchemaParametersFrom(schemaName);
+            SetUniqueName();
+            m_OriginalName = m_InputControlSchemeName;
+        }
+
+        private void SetSchemaParametersFrom(string schemaName)
+        {
+            m_InputControlSchemeName = m_AssetManager.m_AssetObjectForEditing.GetControlScheme(schemaName).name;
+            var schema = m_AssetManager.m_AssetObjectForEditing.GetControlScheme(schemaName);
+            m_Devices = schema.devices.Select(a => new DeviceEntryForList() { name = a.devicePath, deviceEntry = a }).ToList();
+        }
+
+        public override Vector2 GetWindowSize()
+        {
+            return m_ButtonsAndLabelsHeights > 0 ? new Vector2(300, m_ButtonsAndLabelsHeights) : s_DefaultSize;
+        }
+
+        public override void OnOpen()
+        {
+            m_DevicesReorderableList = new ReorderableList(m_Devices, typeof(InputControlScheme.DeviceEntry));
+            m_DevicesReorderableList.headerHeight = 2;
+            m_DevicesReorderableList.onAddCallback += OnDeviceAdd;
+            m_DevicesReorderableList.onRemoveCallback += OnDeviceRemove;
+        }
+
+        void OnDeviceRemove(ReorderableList list)
+        {
+            list.list.RemoveAt(list.index);
+            list.index = -1;
+        }
+
+        void OnDeviceAdd(ReorderableList list)
```

Comment:
No way we can re-use the control picker here and just have an option on it to only show devices and not controls? I'm just thinking that with all the improvements the control picker is still mean to receive, it'd be great to see them here as well. And to also have stuff like filtering here. Plus, I think later we want to have some global preferences that affect what we even consider for the control picker and it'd be great to have that automatically come through here as well.

Reply:
We could but in don't think it makes sense in the current form. When we switch to AdvancedDropdown, the control will have to be (partially) rewritten anyway